### PR TITLE
driver: unmap/flush/FUA support

### PR DESCRIPTION
**Pull request by petrutlucian94**
driver: unmap/flush/FUA support

Those features are already supported by libwnbd and our nbd
client, we'll need to update our IO dispatcher to propagate
them.

Signed-off-by: Lucian Petrut <lpetrut@cloudbasesolutions.com>

### CODE REVIEW

File: libwnbd/wnbd_ioctl.cpp [hunk 1]
```diff
@@ -95,10 +99,17 @@ DWORD WnbdOpenAdapterEx(PHANDLE Handle, PDEVINST CMDeviceInstance)
...
-            "driver is installed. Error: %d. Error message: %s",
-            ErrorCode, win32_strerror(ErrorCode).c_str());
+        if (ErrorCode == ERROR_ACCESS_DENIED) {
+            LogError(
+                "Could not open WNBD adapter device. Access denied, try "
+                "using an elevated command prompt.");
+        }
+        else {
+            LogError(
+                "Could not oped WNBD adapter device. Please make sure that "
```
Comment (aserdean) [inline libwnbd/wnbd_ioctl.cpp:109]: nit: open?

File: driver/wnbd_dispatch.c [hunk 2]
```diff
@@ -190,6 +189,35 @@ NTSTATUS WnbdDispatchRequest(
 
             RtlCopyMemory(Buffer, SrbBuffer, Element->DataLength);
             break;
+        case WnbdReqTypeFlush:
+            Request->Cmd.Flush.BlockAddress =
+                Element->StartingLbn / DevProps->BlockSize;;
```
Comment (aserdean) [inline driver/wnbd_dispatch.c:194]: nit: s/;;/;/

File: libwnbd/wnbd_ioctl.cpp [hunk 3]
```diff
@@ -95,10 +99,17 @@ DWORD WnbdOpenAdapterEx(PHANDLE Handle, PDEVINST CMDeviceInstance)
         *CMDeviceInstance = DevInfoData.DevInst;
     }
     else {
-        LogError(
-            "Could not oped WNBD device. Please make sure that the "
-            "driver is installed. Error: %d. Error message: %s",
-            ErrorCode, win32_strerror(ErrorCode).c_str());
+        if (ErrorCode == ERROR_ACCESS_DENIED) {
+            LogError(
+                "Could not open WNBD adapter device. Access denied, try "
```
Comment (aserdean) [inline libwnbd/wnbd_ioctl.cpp:104]: Should the message be shortened to `Access denied, using an elevated command prompt.`?

File: libwnbd/wnbd_ioctl.cpp [hunk 4]
```diff
@@ -95,10 +99,17 @@ DWORD WnbdOpenAdapterEx(PHANDLE Handle, PDEVINST CMDeviceInstance)
         *CMDeviceInstance = DevInfoData.DevInst;
     }
     else {
-        LogError(
-            "Could not oped WNBD device. Please make sure that the "
-            "driver is installed. Error: %d. Error message: %s",
-            ErrorCode, win32_strerror(ErrorCode).c_str());
+        if (ErrorCode == ERROR_ACCESS_DENIED) {
```
Comment (aserdean) [inline libwnbd/wnbd_ioctl.cpp:102]: `win32_strerror` translates the error code using `FormatMessage` which should provide a meaningful error though.

I understand the need to treat this error differently but I'm not really in favor of it.

**Review by aserdean — CHANGES_REQUESTED** [review]

File: libwnbd/wnbd_ioctl.cpp (hunk 1 above)
Comment (petrutlucian94) [reply libwnbd/wnbd_ioctl.cpp:109]: Whops, thanks.

File: libwnbd/wnbd_ioctl.cpp (hunk 3 above)
Comment (petrutlucian94) [reply libwnbd/wnbd_ioctl.cpp:104]: It's not really the same thing, "Access denied, using an elevated command prompt." suggests that we're automatically using an elevated prompt as a fallback, which is confusing.

File: driver/wnbd_dispatch.c (hunk 2 above)
Comment (petrutlucian94) [reply driver/wnbd_dispatch.c:194]: Whops, thanks.

File: libwnbd/wnbd_ioctl.cpp (hunk 4 above)
Comment (petrutlucian94) [reply libwnbd/wnbd_ioctl.cpp:102]: FormatMessage will just return "Access denied" without any other details. Using unelevated prompts is quite common and most people wouldn't figure out right away the cause of the "Access denied" error, so I really think we should add a more explicit message. There's no downside to it and we'll spare people from some wasted time debugging this, maybe even avoid some bug reports.

File: driver/wnbd_dispatch.c (hunk 2 above)
Comment (aserdean) [reply driver/wnbd_dispatch.c:194]: nit: s/;;/;/g

The first occurrence from `WnbdReqTypeRead` got fixed but there is still one in `WnbdReqTypeFlush`.

**Pull request merged**